### PR TITLE
Factor out shared DAR encoder

### PR DESCRIPTION
**Pull request by cocreature**
changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/archive/encoder/src/main/scala/com/daml/lf/archive/DarWriter.scala [hunk 1]
```diff
@@ -0,0 +1,28 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.lf.archive
+
+import java.io.{ByteArrayOutputStream, OutputStream}
+import java.util.zip.{ZipEntry, ZipOutputStream}
+import scalaz.syntax.traverse._
```
Comment (remyhaemmerle-da) [inline daml-lf/archive/encoder/src/main/scala/com/daml/lf/archive/DarWriter.scala:8]: just wondering where are you using that.

File: daml-lf/archive/encoder/src/main/scala/com/daml/lf/archive/DarWriter.scala [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+object DarWriter {
+  private val manifestPath = "META-INF/MANIFEST.MF"
+
+  def encode(sdkVersion: String, dar: Dar[(String, Array[Byte])], out: OutputStream): Unit = {
```
Comment (remyhaemmerle-da) [inline daml-lf/archive/encoder/src/main/scala/com/daml/lf/archive/DarWriter.scala:13]: You could add a comment, stating that out will be closed.

File: daml-lf/archive/encoder/BUILD.bazel [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    "//bazel_tools:scala.bzl",
+    "da_scala_binary",
+    "da_scala_library",
+    "da_scala_test",
+)
+
+da_scala_library(
+    name = "encoder",
+    srcs = glob(["src/main/scala/**/*.scala"]),
+    main_class = "com.daml.script.dump.Main",
```
Comment (aherrmann-da) [inline daml-lf/archive/encoder/BUILD.bazel:14]:
```suggestion
```
I'm guessing this is a left-over?

File: daml-lf/archive/encoder/src/main/scala/com/daml/lf/archive/DarWriter.scala (hunk 2 above)
Comment (remyhaemmerle-da) [reply daml-lf/archive/encoder/src/main/scala/com/daml/lf/archive/DarWriter.scala:13]: Maybe another version with `out: Path` would be handy.

**Review by aherrmann-da — APPROVED** [review]

**Review by remyhaemmerle-da — APPROVED** [review]
LGTM.

File: daml-lf/archive/encoder/src/main/scala/com/daml/lf/archive/DarWriter.scala (hunk 1 above)
Comment (cocreature) [reply daml-lf/archive/encoder/src/main/scala/com/daml/lf/archive/DarWriter.scala:8]: `Dar.map`

File: daml-lf/archive/encoder/BUILD.bazel (hunk 3 above)
Comment (cocreature) [reply daml-lf/archive/encoder/BUILD.bazel]: absolutely, good catch

File: daml-lf/archive/encoder/src/main/scala/com/daml/lf/archive/DarWriter.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply daml-lf/archive/encoder/src/main/scala/com/daml/lf/archive/DarWriter.scala:8]: :+1:

File: daml-lf/archive/encoder/src/main/scala/com/daml/lf/archive/DarWriter.scala (hunk 2 above)
Comment (cocreature) [reply daml-lf/archive/encoder/src/main/scala/com/daml/lf/archive/DarWriter.scala:13]: good idea on both d21f3ae11c1ad40eff38c757d893209fbed1a375

Comment (remyhaemmerle-da) [reply daml-lf/archive/encoder/src/main/scala/com/daml/lf/archive/DarWriter.scala:24]: Perfect.

**Pull request merged**